### PR TITLE
Add const constructors from Range values

### LINKED ISSUE #4: Add `const` range constructors for `ByteSet`
We should be able to create a `ByteSet` from `a..b` or `a..=b` where `a` and `b` are `u8`, and have it be usable in `const`.

The API should look like:

- `const fn from_range(range: std::ops::Range<u8>) -> ByteSet`
- `const fn from_range_inclusive(range: std::ops::RangeInclusive<u8>) -> ByteSet`

I expect this would be implemented via bit tricks over the integer chunk representation.

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes #4

### CODE REVIEW

File: src/tests.rs [hunk 1]
```diff
@@ -99,3 +99,36 @@ fn last() {
     assert_last_eq!(set, None);
     assert_eq!(set.pop_last(), None);
 }
+
+#[test]
+fn test_open_ranges() {
```
Comment (nvzqz) [inline src/tests.rs]:
```suggestion
fn from_open_ranges() {
```

This name is a bit more clear as to what's being tested.

File: src/tests.rs [hunk 2]
```diff
@@ -99,3 +99,36 @@ fn last() {
...
+        for b in 0..=u8::max_value() {
+            assert_eq!(range_to.contains(&b), set_to.contains(b));
+            assert_eq!(range_from.contains(&b), set_from.contains(b));
+            assert_eq!(range_to_i.contains(&b), set_to_i.contains(b));
+        }
+    }
+}
+
+#[test]
+fn test_closed_ranges() {
```
Comment (nvzqz) [inline src/tests.rs]:
```suggestion
fn from_closed_ranges() {
```
Same reason as `test_open_ranges`.

File: src/lib.rs [hunk 3]
```diff
@@ -1479,6 +1556,49 @@ impl From<&str> for ByteSet {
...
+        Self::from_range_inclusive(s)
+    }
+}
+
+impl From<ops::RangeToInclusive<u8>> for ByteSet {
+    #[inline]
+    fn from(s: ops::RangeToInclusive<u8>) -> Self {
+        Self::from_range_to_inclusive(s)
+    }
+}
```
Comment (nvzqz) [inline src/lib.rs]:
```suggestion
impl From<ops::Range<u8>> for ByteSet {
    #[inline]
    fn from(range: ops::Range<u8>) -> Self {
        Self::from_range(range)
    }
}

impl From<ops::RangeTo<u8>> for ByteSet {
    #[inline]
    fn from(range: ops::RangeTo<u8>) -> Self {
        Self::from_range_to(range)
    }
}

impl From<ops::RangeFrom<u8>> for ByteSet {
    #[inline]
    fn from(range: ops::RangeFrom<u8>) -> Self {
        Self::from_range_from(range)
    }
}

impl From<ops::RangeInclusive<u8>> for ByteSet {
    #[inline]
    fn from(range: ops::RangeInclusive<u8>) -> Self {
        Self::from_range_inclusive(range)
    }
}

impl From<ops::RangeToInclusive<u8>> for ByteSet {
    #[inline]
    fn from(range: ops::RangeToInclusive<u8>) -> Self {
        Self::from_range_to_inclusive(range)
    }
}
```

File: src/lib.rs [hunk 4]
```diff
@@ -540,6 +540,83 @@ impl ByteSet {
...
+            // }
+            //
+            // Once `if` works in const, this can be cleaned up
+            // https://github.com/rust-lang/rust/pull/72437
+            [
+                [
+                    0, Chunk::max_value()
+                ][(this_chunk < byte_chunk) as usize],
+                value
+            ][(this_chunk == byte_chunk) as usize]
```
Comment (nvzqz) [inline src/lib.rs]: Please split the indexing out to two separate operations to make this easier to read.

File: src/lib.rs [hunk 5]
```diff
@@ -540,6 +540,83 @@ impl ByteSet {
...
+        }
+        let (index, shift) = chunk_index_and_shift(range.end);
+        #[cfg(byte_set_chunk_64)]
+        let array = [
+            chunk_for(0, index, shift),
+            chunk_for(1, index, shift),
+            chunk_for(2, index, shift),
+            chunk_for(3, index, shift),
+        ];
+        #[cfg(not(byte_set_chunk_64))]
```
Comment (nvzqz) [inline src/lib.rs]:
```suggestion
        #[cfg(target_pointer_width = "64")]
        let array = [
            chunk_for(0, index, shift),
            chunk_for(1, index, shift),
            chunk_for(2, index, shift),
            chunk_for(3, index, shift),
        ];
        #[cfg(not(target_pointer_width = "64"))]
```

**Review by nvzqz — CHANGES_REQUESTED** [review]
Thanks for this Manish! I just have a few nits but aside from that it looks good to me. 🙂

**Comment by Manishearth** [comment]
Addressed

**Review by nvzqz — APPROVED** [review]
LGTM! Just address the formatting errors and I'll merge 👍

**Comment by Manishearth** [comment]
Done!

**Pull request merged**